README
Wodan Reverse Proxy README
==========================

---------
Contents:
---------

1. About Wodan
2. When to use Wodan
3. When NOT to use Wodan or when to think twice about using it
4. Advantages over other proxy systems
5. Installation
6. Configuration
7. Running
8. Per-page configuration and extra headers
9. Contact Information


--------------
1. About Wodan
--------------

Wodan is a reverse proxy module for the Apache web browser. It caches web
content for better performance. A typical setup would be somewhat like this:

	--------------
	|            |
	| Web client |------\
	|            |       \   ---------    --------------
	--------------        ---|       |    |            |
	                         | Wodan |----| Web Server |
	--------------        ---|       |    |            |
	|            |       /   ---------    --------------
	| Web client |------/
	|            |
	--------------

The web clients don't know anything about the reverse proxy. They get content
just like they would directly from a web server. But the web server gets a lot
less work to do.


--------------------
2. When to use Wodan
--------------------

- The web server uses a content management system that requires a lot of cpu
  power to generate live pages, but the pages don't change too often.
- To hide the web servers from the internet.


---------------------------------------------------------------
3. When NOT to use Wodan, or when to think twice about using it
---------------------------------------------------------------

- The web site has a user login/authentication system. Or rather, do not use
  it if the biggest part of your site is hidden behind a login system. If
  you have a much visited set of pages and a part behind a login system, you
  can still use mod_wodan to cache the public part. 

- The web site site has very dynamic content. In this case, mod_wodan will
  not gain you much in terms of server load. You can still cache things
  like graphics and css files though, so there still might be some gain.

--------------------------------------
4. Advantages over other proxy systems
--------------------------------------

Many advantages of Wodan are a result of the fact that it is an apache module.
Here is an overview of a few important advantages:

- Simple open caching system
- Uses custom headers so you can control caching yourself
- High performance
- Use of other apache modules like usertracking
- The most used other open source reverse proxy, Squid, is too nice to clients.
  Whenever a client really wants the page from the webserver, Squid passes the
  request to the backend. Wodan, however, will always serve from it's cache, unless
  the configuration tells it not to.
  
---------------
5. Installation
---------------

See INSTALL file.	

----------------
6. Configuration
----------------
	
Configuration of Wodan is done in the httpd.conf file of Apache. Here is a list
of directives and what they do. The directives can also be used in VirtualHost
sections:

	----------------------
	WodanPass <dir> <url>
	----------------------

	Map a directory to a remote server, e.g. WodanPass / http://www.wodan.net to map
	the entire host to the wodan website. You can also specify a subdirectory, e.g.
	WodanPassReverse /wodan/ http://www.wodan.net to map the /wodan/ dir of the host to the
	wodan website. The url field may contain port and/or subdir info, e.g. 
	http://www.wodan.net:8080/docs/

	----------------------
	WodanPassReverse <dir> <url>
	----------------------

	Reverse map a remote host to the subdir. This means wodan adjusts some header fields 
	received from the webserver (e.g. Location:). Wodan replaces the <url> part in the headers
	by the hostname and <dir> tag. Only use this tag when also using WodanPassReverse, e.g.:

	WodanPass / http://www.wodan.net:8080/
	WodanPassReverse / http://www.wodan.net/

	NOTE: a lot of webservers don't send port information in the headers, so you shouldn't 
	specify the port of the webserver. You can see this in the example above.

	-----------------------
	WodanDefaultCacheTime <dir> <time>
	-----------------------

	Specifies how long pages in the specified dir should stay in cache before 
	they are refetched from the webserver. If this is not specified it is 1 hour. 
	The syntax for specifying time is:

		<number><letter> | no-cache

	<letter> specifies the unit of time. It can be:
		's'/nothing: seconds
		'm':         minutes
		'h':         hours
		'd':         days
		'w':         weeks

	'no-cache' should be obvious :-)

	-------------------------------------------
	WodanDefaultCacheTimeMatch <regexpattern> <time>
	-------------------------------------------

	Specifies how long pages (or other web-objects) with a uri conforming to regexpattern
	will stay in cache. The syntax for specifying time is the same as with
	the WodanDefaultCacheTime directive. 

	Note that this directive will take precedence over WodanDefaultCacheTime. 
	Because of this, the following directives will have the result of caching
	all files inside the /var/www dir for 2 minutes, except the JPGs, which 
	will be cached for 30 minutes:

	WodanDefaultCacheTime /var/www 2m
	WodanDefaultCacheTimeMatch ^/var/www/.*jpg$ 30m

	---------------------------------------------------------------
	WodanDefaultCacheTimeHeaderMatch <http-header> <regexpattern> <time>
	---------------------------------------------------------------

	Specifies how long web-objects which are send with a HTTP-header
	conforming to the regular expression pattern will stay in cache. The
	syntax for specifying time is the same as with the WodanDefaultCacheTime
	directive.

	Please not that this directive will take precedence over other 
	WodanDefaultCacheTime directives.

	The following directives will cache everything inside the /var/www 
	directory for two minutes, except for all images (all objects with
	a mime-type of image/*, which will be cached for 30 minutes:

	WodanDefaultCacheTime /var/www 2m
	WodanDefaultCacheTimeHeaderMatch Content-Type ^image/.*$ 30m 
    

	-----------------------------
	WodanHashHeader <http-header> 
	-----------------------------

	Specifies additional request headers to use in constructing the hash for 
	a request. 

	The following directives will add support for multi-lingual sites.

	WodanHashHeader Cookie
	WodanHashHeader Accept-Language

	-------------------------------------------------
	WodanHashHeaderMatch <http-header> <regexpattern>
	-------------------------------------------------

	This adds regexp matching on headers, and allows substitions as well:

	for example, use:

	WodanHashHeaderMatch Cookie '.*(I18N_LANGUAGE=[^;]+).*' $1

	to extract the I18N_LANGUAGE cookie and use it in calculating the hash
	for a particular request.

	--------------	
	WodanCacheDir <dir>
	--------------

	The directory in which the cachefiles are stored. Make sure the directory 
	is writable by the user under which Apache runs. For example:

	WodanCacheDir /usr/local/apache_wodan/cache

	To create the cache directory:
      
	# mkdir /usr/local/apache_wodan/cache
	# chown nobody:nobody /usr/local/apache_wodan/cache

	('nobody' is assumed to be the user under which apache runs)

	----------------------
	WodanCacheDirLevels <levels>
	----------------------

	The number of nested subdirectories that will be created in the WodanCacheDir.
	For example, with two levels, a cache file called
	'ce8d3d2136d2528b1f59b23864a4a0ba' will be stored in '<cachedir>/c/e/'.

	The default is two levels, the maximum is 8 levels.

	-----------------
	WodanRunOnCache on|off
	-----------------

	If set to on, the (virtual) server will not use the backend. This can for
	instance be useful if there is some scheduled downtime on the backend, but
	you would like to keep a site on-line.

	If a requested page cannot be found in the cache, a 404 NOT FOUND error 
	is sent to the client.

	The default value for WodanRunOnCache is 'off'.

	----------------
	WodanCache404s yes|no
	----------------

	If set to "yes", Wodan will store 404 status codes in cachefiles
	for URL that get this return code from the backend. Their default expire
	time is the same as for other documents.

	The default value for Cache404s is 'no'.

	------------------------------------
	WodanBackendTimeout <time in milliseconds>
	------------------------------------

	If set, this specifies a timeout for the backend connection. This controls
	the amount of time the reverse proxy waits for a connection with the 
	backend can be made. This includes the time it takes to get the first
	byte of information from the backend. 
	If a connection cannot be made, the page is taken
	from the cache, even if it is supposed to be expired. When the page
	is not in cache, a 404 is generated if the connection times out. 
	When the file is served from cache, the cache is rewritten with
	a new expiry time, so future requests for the same file will
	not go the backend, until the time expires.

	This directive can be useful when a backend is not very stable, but the
	site needs to stay up (albeit with a slight delay, set by <time>)

	The Timeout time is given in millisecond (1/1000 second). e.g. 
	WodanBackendTimeout 1500 is used for a timeout of 1.5 seconds.
	The maximum value for time is 60000 milliseconds (60 seconds). If fed with 
	a higher number, the BackendTimeout will be set to 60000 milliseconds.

	The default value for WodanBackendTimeout is 0, which means that the timeout
	will not be used (only the normal apache and TCP timeouts will be used.

----------
7. Running
----------

Running Wodan is really simple. Just start apache with:

	# <apachedir>/bin/apachectl start

-------------------------------------------
8. Per-page configuration and extra headers
-------------------------------------------

	--------------
	X-Wodan header
	--------------

	In addition to the default cache time values, you can specify 
	cache time values per page. To do this, the web server serving 
	the real pages must add an additional header:

		X-Wodan

	This header can contain the following information:

		'expire <number><letter>' or
		'no-cache'

	The syntax for '<number><letter>' is the same as in the 
	Apache configuration.
	
	Examples:

		X-Wodan: no-cache
		X-Wodan: expire 35m

	If you are using PHP in your pages, it's very easy to send 
	these extra headers. Just put the following code at the top of 
	your .php file (this has to be before any HTML content, 
	because headers have to be sent before that):

		<?php header("X-Wodan: expire 10m"); ?>

----------------------
9. Logging
----------------------

Wodan places an Apache server note called "WodanSource" with every request. This
note can have the following values:
* Cached: the returned page comes from cache, as planned
* Backend: the returned page comes from the backend
* CachedBackendError: the returned page comes from cache. It should have been 
  gotten from the backend, but an error has occurred while getting it from this
  backend.

The value of this note can be used for logging. If you add
%{WodanSource}n the your logformat, your TransferLog will hold the value
of the WodanSource header.

The other note that is set is "WodanBackendTime". This note is set for every 
request that gets it's content from the backend. It records the time that 
it takes to get the whole request from the backend. To get this into
your access log, add %{WodanBackendTime} to your logformat.

-----------------------
10. Tools
-----------------------

The tools/ directory holds tools that can be used for maintaining
a system with Wodan.

1. remove_expired.py

This very simple Python script removed all expired cache files
in a directory. It does this recursively.

e.g.:
remove_expired.py /var/wodan/cache 

will remove all expired cache files under that directory.

The best way to use this script is to run it using cron.

-----------------------
11. Contact Information
-----------------------

Wodan was created by IC&S (http://www.ic-s.nl/), a Dutch Linux company, and of
course remains under active development.

Website      : http://www.wodan.net/
E-mail       : info@wodan.net
Mailing List : wodan@wodan.net
               http://twister.fastxs.net/mailman/listinfo/wodan
               http://twister.fastxs.net/pipermail/wodan/



debian/changelog
libapache2-mod-wodan (2.2.0rc1-1) unstable; urgency=low

  * Initial upload.

 -- Paul J Stevens <paul@nfg.nl>  Sun, 18 Oct 2009 16:41:22 +0200


mod_wodan.c
/* 
 * (c) 2000-2006 IC&S, The Netherlands
 */ 

#define WODAN_NAME "Wodan"
#define WODAN_VERSION "2.0.1"

/* constants identifying the source of the returned (to the client) object */
#define LOG_SOURCE_CACHED "Cached"
#define LOG_SOURCE_BACKEND "Backend"
#define LOG_SOURCE_CACHED_BACKEND_ERROR "CachedBackendError"

/* local includes */
#include "cache.h"
#include "datatypes.h"
#include "httpclient.h"
#include "match.h"
#include "util.h"

/* Apache includes */
#include "httpd.h"
#include "http_config.h"
#include "http_log.h"
#include "http_protocol.h"
#include "ap_config.h"
#include "apr_strings.h"
#include <string.h>
#include <time.h>

/*
 * Function prototypes.
 */
 
/* initializer for Wodan */
static int wodan_init_handler(apr_pool_t *p, apr_pool_t *plog, apr_pool_t *ptemp,
	server_rec *s);
/* create initial server configuration. */
static void *wodan_create_server_config(apr_pool_t *p, server_rec *s);
/* create per-dir server configuration */
static void *wodan_create_dir_config(apr_pool_t *p, char *dir);
/* merge two config structures. */
static void *wodan_merge_config(apr_pool_t *p, void *base_config_p, 
	void *new_config_p);
/* add a WodanPass config to the configuration */
static const char *add_pass(cmd_parms *cmd, void *dummy, const char *path,
	const char *url);
/* add a WodanPassReverse config to the configuration */
static const char *add_pass_reverse(cmd_parms *cmd, void *dummy, const char *path,
	const char *url);
/* add cachedir */	
static const char *add_cachedir(cmd_parms *cmd, void *dummy, const char *path);
/* set level of cachedir nesting */
static const char *add_cachedir_levels(cmd_parms *cmd, void *dummy, 
	const char *level);
/* add a default cachetime for a path */
static const char *add_default_cachetime(cmd_parms *cmd, void *dummy,
	const char *path, const char *time_string);
/* add a default cachetime for a path matching a regular expression */
static const char* add_default_cachetime_regex(cmd_parms *cmd, 
	void *dummy, const char *regex_pattern, const char *time_string);
/* add a header to the hash */
static const char* add_hash_header(cmd_parms *cmd, 
	void *dummy, const char *header);
/* add a header to the hash if the value matches the regex pattern */
static const char* add_hash_header_match(cmd_parms *cmd, 
	void *dummy, const char *header, const char *regex_pattern, const char *replacement);
/* add a default cachetime for a header whose value matches a regular expression */
static const char* add_default_cachetime_header(cmd_parms *cmd, void *dummy, 
	const char *http_header, const char *regex_pattern, const char *time_string);
/* add a flag to run completely on the cache (i.e. do not try to contact the 
 * backend */
static const char* add_run_on_cache(cmd_parms *cmd, void *dummy, int flag);
/* add a flag to cache 404 pages. */
static const char* add_cache_404s(cmd_parms *cmd, void *dummy, int flag);
/* add a backend timeout */
static const char *add_backend_timeout(cmd_parms *cmd, void *dummy,
	const char *timeout_string);

/* hook registering function */
static void wodan_register_hooks(apr_pool_t *p);
	
/* The content handler function. This is the main function of Wodan */
static int wodan_handler(request_rec *r);

/**
 * The configuration directives and their setup functions.
 */
static const command_rec wodan_commands[] = 
{
	AP_INIT_TAKE12("WodanPass", add_pass, NULL, RSRC_CONF, "A path and a URL"),
	AP_INIT_TAKE12("WodanPassReverse", add_pass_reverse, 
		NULL, RSRC_CONF, "A path and a URL"),
	AP_INIT_TAKE1("WodanCacheDir", add_cachedir, NULL, RSRC_CONF, "A path"),
	AP_INIT_TAKE1("WodanCacheDirLevels", add_cachedir_levels, NULL, RSRC_CONF, 
		"A Number (> 0)"),
	AP_INIT_TAKE2("WodanDefaultCacheTime", add_default_cachetime, NULL, RSRC_CONF,
		"A path and a time string"),
	AP_INIT_TAKE2("WodanDefaultCacheTimeMatch", add_default_cachetime_regex,
		NULL, RSRC_CONF, "A regex pattern and a time string"),
	AP_INIT_TAKE3("WodanDefaultCacheTimeHeaderMatch", 
		add_default_cachetime_header, NULL, RSRC_CONF, 
		"A header, a regex pattern and a time string"),
	AP_INIT_TAKE1("WodanHashHeader",
		add_hash_header, NULL, RSRC_CONF,
		"A header to add to the hash"),
	AP_INIT_TAKE23("WodanHashHeaderMatch",
		add_hash_header_match, NULL, RSRC_CONF,
		"A header to add to the hash if the regex pattern matches and an optional replacement pattern"),
	AP_INIT_FLAG("WodanRunOnCache", add_run_on_cache, NULL, RSRC_CONF,
		"run completely on cache"),
	AP_INIT_FLAG("WodanCache404s", add_cache_404s, NULL, RSRC_CONF,
		"cache 404 pages"),
	AP_INIT_TAKE1("WodanBackendTimeout", add_backend_timeout, NULL, RSRC_CONF,
		"a number, which represents a time in miliseconds"),
	{NULL}
};
/* The module. Apache uses this information to initialise and hook up
 * the module into the webserver. */
module AP_MODULE_DECLARE_DATA wodan_module = {
    STANDARD20_MODULE_STUFF, 
    wodan_create_dir_config,   /* create per-dir    config structures */
    wodan_merge_config,        /* merge  per-dir    config structures */
    wodan_create_server_config,/* create per-server config structures */
    wodan_merge_config,        /* merge  per-server config structures */
    wodan_commands,            /* table of config file commands       */
    wodan_register_hooks       /* register hooks                      */
};

/* initialize Wodan */
static int wodan_init_handler(apr_pool_t *p, 
	apr_pool_t *plog UNUSED, 
	apr_pool_t *ptemp UNUSED,
	server_rec *s UNUSED)
{
	const char *identifier_string;
	
	identifier_string = apr_psprintf(p, "%s/%s", WODAN_NAME, WODAN_VERSION);
	ap_add_version_component(p, identifier_string);
	
	return OK;
}                          

static void *wodan_create_config(apr_pool_t *p)
{
	wodan_config_t* config = (wodan_config_t *) apr_pcalloc(p, sizeof(wodan_config_t));
	
	config->cachedir_levels           = DEFAULT_CACHEDIR_LEVELS;
	config->proxy_passes              = apr_array_make(p, 0, sizeof(wodan_proxy_destination_t));
	config->proxy_passes_reverse      = apr_array_make(p, 0, sizeof(wodan_proxy_alias_t));
	config->default_cachetimes        = apr_array_make(p, 0, sizeof(wodan_default_cachetime_t));
	config->default_cachetimes_regex  = apr_array_make(p, 0, sizeof(wodan_default_cachetime_regex_t));
	config->default_cachetimes_header = apr_array_make(p, 0, sizeof(wodan_default_cachetime_header_t));	
	config->hash_headers              = apr_array_make(p, 0, sizeof(const char *));
	config->hash_headers_match        = apr_array_make(p, 0, sizeof(wodan_hash_header_match_t));
	return config;		
}

static void *wodan_create_server_config(apr_pool_t *p, 
	server_rec *s UNUSED)
{
	return wodan_create_config(p);
}

static void *wodan_create_dir_config(apr_pool_t *p, 
	char *dir UNUSED)
{
	return wodan_create_config(p);
}
	          
static void *wodan_merge_config(apr_pool_t *p, void *base_config_p, 
	void *new_config_p)
{
	wodan_config_t *config = (wodan_config_t *)
		apr_pcalloc(p, sizeof(wodan_config_t));
	wodan_config_t *base_config = (wodan_config_t *) base_config_p;
	wodan_config_t *new_config = (wodan_config_t *) new_config_p;
	
	if (strlen(new_config->cachedir) > 0) 
		apr_cpystrn(config->cachedir, new_config->cachedir, 
			MAX_CACHE_PATH_SIZE + 1);
	else 
		apr_cpystrn(config->cachedir, base_config->cachedir,
			MAX_CACHE_PATH_SIZE + 1);
	
	config->cachedir_levels = new_config->cachedir_levels;
	if (new_config->is_cachedir_set == 1 || base_config->is_cachedir_set == 1)
		config->is_cachedir_set = 1;
	if (new_config->run_on_cache == 1 || base_config->run_on_cache == 1)
		config->run_on_cache = 1;
	if (new_config->cache_404s == 1 || base_config->cache_404s == 1)
		config->cache_404s = 1;
	if (new_config->backend_timeout != (apr_interval_time_t) 0)
		config->backend_timeout = new_config->backend_timeout;
	else
		config->backend_timeout = base_config->backend_timeout;
		
	config->proxy_passes = apr_array_append(p, 
		base_config->proxy_passes, new_config->proxy_passes);
	config->proxy_passes_reverse = apr_array_append(p,
		base_config->proxy_passes_reverse, 
		new_config->proxy_passes_reverse);
	config->default_cachetimes = apr_array_append(p,
		base_config->default_cachetimes, new_config->default_cachetimes);
	config->default_cachetimes_regex = apr_array_append(p,
		base_config->default_cachetimes_regex, 
		new_config->default_cachetimes_regex);
	config->default_cachetimes_header = apr_array_append(p,
		base_config->default_cachetimes_header, 
		new_config->default_cachetimes_header);
	config->hash_headers = apr_array_append(p,
		base_config->hash_headers, 
		new_config->hash_headers);
	config->hash_headers_match = apr_array_append(p,
		base_config->hash_headers_match, 
		new_config->hash_headers_match);

	return config;
}               
/* The sample content handler */

static const char *add_pass(cmd_parms *cmd, void *dummy UNUSED, 
	const char *path, const char *url)
{
	server_rec *s = cmd->server;
	wodan_config_t *config = (wodan_config_t *)
		ap_get_module_config(s->module_config, &wodan_module);
	wodan_proxy_destination_t *proxy_destination;
	char *proxy_url;
	
	if(path[0] != '/' || path[(int) strlen(path) - 1] != '/')
	        return "First argument of WodanPass should be a dir e.g. /dir/";
	if (strncasecmp(url, "http://", 7) != 0) 
		return "Second argument of WodanPass should be a "
			"http:// url, e.g. http://www.ic-s.nl";
	
	proxy_url = apr_pstrdup(cmd->pool, url);
	/* strip final '/' of proxy_url */
	if (proxy_url[(int) strlen(proxy_url) - 1] == '/')
		proxy_url[(int) strlen(proxy_url) - 1] = '\0';
	
	proxy_destination = apr_array_push(config->proxy_passes);
	proxy_destination->path = path;
	proxy_destination->url = proxy_url;
	return NULL;
}

static const char *add_pass_reverse(cmd_parms *cmd,
	void *dummy UNUSED, const char *path, const char *url)
{
	server_rec *s = cmd->server;
	wodan_config_t *config = (wodan_config_t *)
		ap_get_module_config(s->module_config, &wodan_module);
	wodan_proxy_destination_t *proxy_alias;
	
	if(path[0] != '/' || path[(int) strlen(path) - 1] != '/')
	        return "First argument of WodanPassReverse should be a dir e.g. /dir/";
	if (strncasecmp(url, "http://", 7) != 0) 
		return "Second argument of WodanPassReverse should be a "
			"http:// url, e.g. http://www.ic-s.nl";
	
	proxy_alias = apr_array_push(config->proxy_passes_reverse);
	proxy_alias->path = path;
	proxy_alias->url = url;
	return NULL;
}

static const char *add_cachedir(cmd_parms *cmd, void *dummy UNUSED, 
	const char *path)
{
	server_rec *s = cmd->server;
	wodan_config_t *config = (wodan_config_t *)
		ap_get_module_config(s->module_config, &wodan_module);
	
	/* prepend server root path */
	const char *fname = ap_server_root_relative(cmd->pool, path);
	
	if (!ap_is_directory(cmd->pool, fname)) {
		char *error_message = apr_psprintf(cmd->pool, 
			"WodanCacheDir %s is not a directory!", fname);
		return error_message;
	}
	
	if (!util_file_is_writable(cmd->pool, fname))
	{
		
		char *error_message = apr_psprintf(cmd->pool,
			"WodanCachedir %s should be owned by Wodan user and should be writable!"
			"by that user!", fname);
		return error_message;
	}
	
	apr_cpystrn(config->cachedir, fname, MAX_CACHE_PATH_SIZE + 1);
	config->is_cachedir_set = 1;
	
	return NULL;
}	

static const char *add_cachedir_levels(cmd_parms *cmd, 
	void *dummy UNUSED, const char *level)
{
	server_rec *s = cmd->server;
	wodan_config_t *config = (wodan_config_t *)
		ap_get_module_config(s->module_config, &wodan_module);
	apr_int64_t levels;

	if (!util_string_is_number(level)) {
		char *error_message = apr_psprintf(cmd->pool, 
			"Argument to WodanCacheDirLevels should be a number, it is %s now",
			level);
		return error_message;
	}
	
	levels = apr_strtoi64(level, NULL, 10);
	if (levels < 0 || levels > (apr_int64_t) MAX_CACHEDIR_LEVELS) {
		char *error_message = apr_psprintf(cmd->pool, 
			"WodanCacheDirLevels must have a value between 0 and %d",
			(int) MAX_CACHEDIR_LEVELS);
		return error_message;
	}
	
	config->cachedir_levels = (int) levels;
	return NULL;
}	

static const char *add_default_cachetime(cmd_parms *cmd, 
	void *dummy UNUSED, const char *path, const char *time_string)
{
	server_rec *s = cmd->server;
	wodan_config_t *config = (wodan_config_t *)
		ap_get_module_config(s->module_config, &wodan_module);
	wodan_default_cachetime_t *new_default_cachetime;
	
	if(path[0] != '/' || path[(int) strlen(path) - 1] != '/')
		return "First argument of WodanDefaultCacheTime "
	    		"should be a path, e.g. /dir/";
	 
	new_default_cachetime = apr_array_push(config->default_cachetimes);
	new_default_cachetime->path = path;
	if (strncmp(time_string, "no", 2 ) == 0 )
		new_default_cachetime->cachetime = (apr_int32_t) -1;
	else { 
	 	new_default_cachetime->cachetime = 
	 		util_timestring_to_seconds(apr_pstrdup(cmd->pool, time_string));
	}
	return NULL;
}

static const char* add_default_cachetime_regex(cmd_parms *cmd, 
	void *dummy UNUSED, const char *regex_pattern, 
	const char *time_string)
{
	server_rec *s = cmd->server;
	wodan_config_t *config = (wodan_config_t *)
		ap_get_module_config(s->module_config, &wodan_module);
	wodan_default_cachetime_regex_t *new_default_cachetime_regex;
	ap_regex_t *compiled_pattern = NULL;

	new_default_cachetime_regex = 
		apr_array_push(config->default_cachetimes_regex);
	
	compiled_pattern = ap_pregcomp(cmd->pool, regex_pattern, 
		AP_REG_EXTENDED | AP_REG_NOSUB);
	if (compiled_pattern == NULL) {
		char *error_message = apr_psprintf(cmd->pool, 
			"Failure compiling regex pattern \"%s\"", regex_pattern);
		return error_message;
	}	
	new_default_cachetime_regex->uri_pattern = compiled_pattern;
	
	if (strncmp(time_string, "no", 2) == 0)
		new_default_cachetime_regex->cachetime = (apr_int32_t) -1;
	else
		new_default_cachetime_regex->cachetime = 
			util_timestring_to_seconds(apr_pstrdup(cmd->pool, time_string));

	return NULL;
}

static const char* add_hash_header(cmd_parms *cmd, 
	void *dummy UNUSED, const char *headername)
{
	server_rec *s = cmd->server;
	wodan_config_t *config = (wodan_config_t *) ap_get_module_config(s->module_config, &wodan_module);

	*(const char **)apr_array_push(config->hash_headers) = headername;

	return NULL;
}

static const char* add_hash_header_match(cmd_parms *cmd, 
	void *dummy UNUSED, const char *headername, const char *regex_pattern, const char *replacement)
{
	server_rec *s = cmd->server;
	wodan_config_t *config = (wodan_config_t *) ap_get_module_config(s->module_config, &wodan_module);
	ap_regex_t *compiled_pattern = NULL;
	wodan_hash_header_match_t *directive;

	compiled_pattern = ap_pregcomp(cmd->pool, regex_pattern, AP_REG_EXTENDED);
	if (compiled_pattern == NULL) {
		char *error_message = apr_psprintf(cmd->pool, 
			"Failure compiling regex pattern \"%s\"", regex_pattern);
		return error_message;
	}	

	directive = apr_array_push(config->hash_headers_match);
	directive->header = apr_pstrdup(cmd->pool, headername);
	directive->regex = compiled_pattern;
	directive->pattern = apr_pstrdup(cmd->pool, replacement);
	
	return NULL;
}

static const char* add_default_cachetime_header(cmd_parms *cmd, 
	void *dummy UNUSED, const char *http_header, 
	const char *regex_pattern, const char *time_string)
{
	server_rec *s = cmd->server;
	wodan_config_t *config = (wodan_config_t *)
		ap_get_module_config(s->module_config, &wodan_module);
	wodan_default_cachetime_header_t *new_default_cachetime_header;
	ap_regex_t *compiled_pattern;
	
	new_default_cachetime_header = 
		apr_array_push(config->default_cachetimes_header);
	
	new_default_cachetime_header->header = apr_pstrdup(cmd->pool, http_header);
	compiled_pattern = ap_pregcomp(cmd->pool, regex_pattern, 
		AP_REG_EXTENDED | AP_REG_NOSUB);
	if (compiled_pattern == NULL) {
		char *error_message = apr_psprintf(cmd->pool, 
			"Failure compiling regex pattern \"%s\"", regex_pattern);
		return error_message;
	}
	new_default_cachetime_header->header_value_pattern = compiled_pattern;
	if (strncmp(time_string, "no", 2) == 0) 
		new_default_cachetime_header->cachetime = (apr_int32_t) -1;
	else
		new_default_cachetime_header->cachetime = 
			util_timestring_to_seconds(apr_pstrdup(cmd->pool, time_string));
			
	return NULL;
}

static const char* add_run_on_cache(cmd_parms *cmd, 
	void *dummy UNUSED, int flag)
{
	wodan_config_t *config = (wodan_config_t *)
		ap_get_module_config(cmd->server->module_config, &wodan_module);
      
	config->run_on_cache = (unsigned) flag;
	
	return NULL;
}

static const char *add_cache_404s(cmd_parms *cmd, 
	void *dummy UNUSED, int flag)
{
	wodan_config_t *config = (wodan_config_t *)
		ap_get_module_config(cmd->server->module_config, &wodan_module);
	
	config->cache_404s = (unsigned) flag;
	
	return NULL;
}

static const char *add_backend_timeout(cmd_parms *cmd,
	void *dummy UNUSED, const char *timeout_string)
{
	wodan_config_t *config = (wodan_config_t *)
		ap_get_module_config(cmd->server->module_config, &wodan_module);
	apr_int64_t timeout;
	
	if (!util_string_is_number(timeout_string)) {
		char *error_message = apr_psprintf(cmd->pool,
			"argument should be number, it is \"%s\" now", timeout_string);
		return error_message;
	}
	timeout = apr_strtoi64(timeout_string, NULL, 10);
	
	// timeout is a number in milliseconds, so it needs to be multiplied by 1000
	timeout *= 1000;
	
	if (timeout > apr_time_from_sec(MAX_BACKEND_TIMEOUT_SEC))
		config->backend_timeout = apr_time_from_sec(MAX_BACKEND_TIMEOUT_SEC);
	else
		config->backend_timeout = timeout;
	
	return NULL;
}

static void wodan_register_hooks(apr_pool_t *p UNUSED)
{
	ap_hook_post_config(wodan_init_handler, NULL, NULL, APR_HOOK_MIDDLE);
	ap_hook_handler(wodan_handler, NULL, NULL, APR_HOOK_MIDDLE);
}

static int wodan_handler(request_rec *r)
{
	wodan_config_t* config;
	httpresponse_t httpresponse;
	WodanCacheStatus_t cache_status;
	int response = HTTP_BAD_GATEWAY;
	apr_time_t cache_file_time;
	
	config = (wodan_config_t *)
		ap_get_module_config(r->server->module_config, &wodan_module);
	
	// TODO: check if is is perhaps a better idea to create a table of a certain size
	memset(&httpresponse, 0, sizeof(httpresponse_t));
	httpresponse.headers = apr_table_make(r->pool, 0);
	
	ap_log_error(APLOG_MARK, APLOG_NOERRNO|APLOG_DEBUG, 0, r->server, 
		     "Processing new request: %s%s", r->hostname, r->unparsed_uri);

	// see if the request can be handled from the cache.
	cache_status = cache_get_status(config, r, &cache_file_time);

	ap_log_error(APLOG_MARK, APLOG_NOERRNO|APLOG_DEBUG, 0, r->server, 
		     "Cache status: %d", cache_status);

	if (config->cache_404s) {
		if (cache_status == WODAN_CACHE_404) {
			ap_log_error(APLOG_MARK, APLOG_NOERRNO|APLOG_DEBUG, 0,
			             r->server, "URL is cached as 404");
			return HTTP_NOT_FOUND;
		}
	}
	
	if (cache_status != WODAN_CACHE_PRESENT) {
		/* attempt to get data from backend */
		wodan_proxy_destination_t *proxy_destination =
			destination_longest_match(config, r->unparsed_uri);

		if(proxy_destination != NULL)//FIXME what if destination is NULL
		{
			char* newpath;
			int l = (int) strlen(proxy_destination->path);
			newpath = &(r->unparsed_uri[l - 1]);
			
			ap_log_error(APLOG_MARK, APLOG_NOERRNO|APLOG_DEBUG, 0,
				     r->server, 
				     "No cache, getting content from remote "
				     "url: %s path: %s", 
				     proxy_destination->url, newpath);
			
			//Get the httpresponse from remote server	
			response = http_proxy(config, proxy_destination->url, newpath, 
					      &httpresponse, r, cache_file_time);
			/* If 404 are to be cached, then already return
			 * default 404 page here in case of a 404. */
			if (config->cache_404s)
				if (response == HTTP_NOT_FOUND)
					return HTTP_NOT_FOUND;

			/* if nothing can be received from backend, and
			   nothing in cache, NOT_FOUND is the only option
			   left... */
			if ((response == HTTP_NOT_FOUND || response == HTTP_BAD_GATEWAY ||
					response == HTTP_GATEWAY_TIME_OUT) &&
					cache_status != WODAN_CACHE_PRESENT_EXPIRED) {
			    	ap_log_error(APLOG_MARK, APLOG_NOERRNO|APLOG_DEBUG, 0, r->server,
			    		"return HTTP_NOT_FOUND");
			    	return HTTP_NOT_FOUND;
			} else if (response != HTTP_BAD_GATEWAY &&
				 response != HTTP_GATEWAY_TIME_OUT &&
				 response != HTTP_NOT_MODIFIED) {
				ap_log_error(APLOG_MARK, 
					     APLOG_NOERRNO|APLOG_DEBUG, 0, r->server,
					     "Got response from gateway");
			} 
		}
	}

	if (cache_status == WODAN_CACHE_PRESENT) {
	  	cache_read_from_cache(config, r, &httpresponse);
		apr_table_set(r->notes, "WodanSource", LOG_SOURCE_CACHED);
	} else if (cache_status == WODAN_CACHE_PRESENT_EXPIRED &&
		   (response == HTTP_BAD_GATEWAY || 
		    response == HTTP_GATEWAY_TIME_OUT ||
		    response == HTTP_NOT_MODIFIED)) {
	  	cache_read_from_cache(config, r, &httpresponse);
		cache_update_expiry_time(config, r);
		apr_table_set(r->notes, "WodanSource", LOG_SOURCE_CACHED_BACKEND_ERROR);
	} else {
		apr_table_set(r->notes, "WodanSource", LOG_SOURCE_BACKEND);
	}

	//Return some response code
	ap_log_error(APLOG_MARK, APLOG_NOERRNO|APLOG_DEBUG, 0, r->server, 
		     "returning: %d",  httpresponse.response);
	
	return OK; 
}




